### PR TITLE
API-12145 F: V1: Add parent facility reference to Facilities records

### PR DESCRIPTION
**Pull request by giangnguyen2**
Add parent reference to V1 facility Record
[API-12145](https://vajira.max.gov/browse/API-12145)

### CODE REVIEW

**Review by mmccollum-va — APPROVED** [review]

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/HealthTransformer.java [hunk 1]
```diff
@@ -264,6 +266,13 @@ private String id() {
     return "vha_" + vast.stationNumber();
   }
 
+  private String parentId() {
```
Comment (ChrisBratti-VA) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/HealthTransformer.java:269]: Could simplify this to a one-liner with a ternary operator:
`return allBlank(vast.parentStationNumber()) ? null : "vha_" + vast.parentStationNumber();`

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/VetCenterTransformer.java [hunk 2]
```diff
@@ -130,6 +132,13 @@ private String id() {
     return "vc_" + vast.stationNumber();
   }
 
+  private String parentId() {
```
Comment (ChrisBratti-VA) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/VetCenterTransformer.java:135]: Could simplify this to a one-liner with a ternary operator:
`return allBlank(vast.parentStationNumber()) ? null : "vc_" + vast.parentStationNumber();`

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/HealthTransformer.java (hunk 1 above)
Comment (giangnguyen2) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/HealthTransformer.java:269]: Good call! This has been addressed

**Review by ChrisBratti-VA — APPROVED** [review]

**Review by mmccollum-va — APPROVED** [review]

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/v1/Facility.java [hunk 3]
```diff
@@ -194,28 +194,6 @@ public boolean isEmpty() {
   @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
   @JsonInclude(value = Include.NON_EMPTY, content = Include.NON_EMPTY)
   @JsonSerialize(using = FacilityAttributesSerializer.class)
-  @JsonPropertyOrder({
```
Comment (as-va) [inline facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/v1/Facility.java:193]: Why was this removed? This defines the ordering of fields in the API response.

File: facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java [hunk 4]
```diff
@@ -298,6 +299,12 @@ private DatamartFacility datamartFacility() {
   public void datamartFacilityRoundtrip() {
     DatamartFacility datamartFacility = datamartFacility();
     Facility facility = FacilityTransformerV1.toFacility(datamartFacility);
+    facility
+        .attributes()
+        .parent(
+            FacilityTransformerV1.toFacilityParent(
+                datamartFacility.attributes().parentId(),
+                "http://localhost:8085/v1/facilities/vha_123"));
```
Comment (as-va) [inline facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java:307]: These tests and resource files aren't just run on our local machines - they are also run on each of the AWS environments during application deployments. This hardcoded address will only work locally; it needs to run everywhere.

Take a look at `SystemDefinitions` in the facilities/test project for a helper class that can figure out where it is running and return the appropriate settings.

**Review by as-va — CHANGES_REQUESTED** [review]

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/v1/Facility.java (hunk 3 above)
Comment (giangnguyen2) [reply facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/v1/Facility.java:193]: For V1, we now use a custom serializer `FacilityAttributesSerializer` which dictates the ordering of fields in the response so this ` @JsonPropertyOrde` code no longer have effect on the ordering of fields

File: facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java (hunk 4 above)
Comment (giangnguyen2) [reply facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java]: Thanks for the reference. I have updated the code.

File: facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java [hunk 5]
```diff
@@ -421,20 +438,39 @@ private Facility facility() {
...
+    DatamartFacility df = FacilityTransformerV1.toVersionAgnostic(facility);
+    assertThat(df).hasFieldOrProperty("attributes.parentId");
+    // Assert that there is no parent for V0 Facility when it is transformed from datamartFacility
+    assertThatThrownBy(
+            () ->
+                assertThat(
+                        FacilityTransformerV0.toFacility(
+                            FacilityTransformerV1.toVersionAgnostic(facility)))
+                    .hasFieldOrProperty("attributes.parent"))
+        .isInstanceOf(AssertionError.class);
```
Comment (mmccollum-va) [inline facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java:466]: Consider adding `.hasMessage(<error_msg_thrown>)` to include verification of the error message thrown in the test

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/DatamartFacility.java [hunk 6]
```diff
@@ -129,27 +128,6 @@
...
-    "operational_hours_special_instructions",
-    "services",
-    "satisfaction",
-    "wait_times",
-    "mobile",
-    "active_status",
-    "operating_status",
-    "detailed_services",
-    "visn"
-  })
```
Comment (mmccollum-va) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/DatamartFacility.java:152]: No `@JsonSerializer` has been specified for `DatamartFacility.FacilityAttributes`. Consider reincluding `@JsonPropertyOrder`

**Review by mmccollum-va — CHANGES_REQUESTED** [review]

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/DatamartFacility.java (hunk 6 above)
Comment (giangnguyen2) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/DatamartFacility.java:152]: I see that there is a custom `FacilityAttributesSerializer` in `DatamartJacksonSerializers`

File: facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java (hunk 5 above)
Comment (giangnguyen2) [reply facilities/src/test/java/gov/va/api/lighthouse/facilities/FacilityTransformerV1Test.java]: I have added a message to this assert.

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/DatamartFacility.java (hunk 6 above)
Comment (mmccollum-va) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/DatamartFacility.java:334]: True, `FacilityAttributesSerializer` in `DatamartJacksonSerializers` does ensure this property order, but it is possible to serialize `DatamartFacility` to JSON outside of `DatamartJacksonSerializers`. Setting `@JsonPropertyOrder` will ensure the expected order in such cases.

**Review by mmccollum-va — CHANGES_REQUESTED** [review]

**Review by mmccollum-va — APPROVED** [review]

**Review by mmccollum-va — DISMISSED** [review]

**Review by ChrisBratti-VA — APPROVED** [review]

**Pull request closed without merging**